doc/vrrp/VRRP_Adaptation_HLD.md

```diff
@@ -0,0 +1,712 @@
+
+IPv4 case: 00-00-5e-00-01-{vrid}
+
+IPv6 case: 00-00-5e-00-02-{vrid}
+
+where, vrid is user configured 1-byte virtual router identifier. VRID has interface scope; that is, VRID has to be unique among the VRRP instances on an interface.
+
+##### VRRP Advertisement Frame
+
+VRRP control packets have IP protocol type as 112 (reserved for VRRP), for IPv4 network, are sent to VRRP multicast address 224.0.0.18, source MAC in VRRP control packets is virtual MAC address and source IP is interface IP (not the virtual IPv4 address). For IPv6 networks, the source address in the VRRPv6 header is the IPv6 link local address of the sending interface, and the destination address is the IPv6 multicast address FF02:: 02/64 assigned by IANA.
+
+##### Uplink Interface Tracking
+
+The fault point which VRRP safeguards against is the interface on which VRRP instance is present. Even if all of the connectivity of master to the external network fails, VRRP will still not trigger the failover to the backup gateway. Uplink interface tracking feature has been introduced to overcome this limitation.
+
+![VRRP Tracking Interface Scenarios](images/VRRP_Tracking_Interface_Scenarios_Diagram.png "Figure 1: VRRP Tracking Interface Scenarios")
+
+As shown in the figure，Configure VRRP backup groups between RouterA and RouterB, Both RouterA and RouterB operate in preemptive mode. On RouterA, configure to monitor the uplink interface Interface1. When Interface1 fails, RouterA lowers its priority and, through packet negotiation, RouterB preempts and becomes the Master, ensuring that user traffic is forwarded normally.
+
+### Architecture Design 
+
+Below diagram illuminates the interactions between relevant components and the DB involved for VRRP support in SONiC architecture at a high level.
+
+![VRRP Architecture](images/VRRP_Architecture_Diagram.png "Figure 2: VRRP in SONiC Architecture")
+
+Diagram 1. VRRP components.
+
+Significantly, This feature extends functionality implemented in [SONiC FRR-BGP Extended Unified Configuration Management Framework](https://github.com/sonic-net/SONiC/blob/master/doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md), Redis DB events will trigger frrcfgd when the field frr_mgmt_framework_config set to "true" in the DEVICE_METADATA table, otherwise will trigger vrrpcfgd, and then frrcfgd or vrrpcfgd will configure FRR-VRRP using FRR CLI commands.
+
+### High-Level Design
+
+#### Operating environment
+
+Since support for placing macvlan devices into protodown was not added to Linux until version 5.1, the kernel version is required:
+- Linux kernel version 5.1+
+
+#### Scalability Requirements
+
+- Max number of VRRP instances: 254
+
+#### Container
+
+##### BGP container
+
+vrrpcfgd: 
+  - Subscribes to CONFIG_DB tables, parsing configurations and passes to vrrpd by using FRR CLI commands. 
+
+vrrpd: 
+  - Responsible for all VRRP protocol related calculations. VRRP packets are sent and received in vrrpd and states are updated accordingly.
```

Comment:
Per FRR documentation,
_FRR’s VRRP implementation uses Linux macvlan devices to to implement the shared virtual MAC feature of the protocol. Currently, it does not create those system interfaces - they must be configured outside of FRR before VRRP can be enabled on them._

Do you know whether FRR has any plan to create MACVLAN interface for VRRP in the near future?

Do we have any sequence issue that VRRP can be enabled in FRR only after the MACVLAN interface creation in the kernel?

Reply:
Have not found such plan, from the vrrpd user guider, need the user to do the creation of Macvlan device for different OS, since it depends on the kernel tool such as iptable2, some OS may not support that.
And  we have not faced such an issue about the sequence currently, and we will pay more attention about it later.